feat: Add formEnctype support in data provider to force multipart encoding

ericovasconcelos opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | main for features / current stable version branch for bug fixes <!-- see below -->
| Tickets       | #387 
| License       | MIT
| Doc PR        | api-platform/docs#... <!-- required for new features -->
<!--
This pull request adds a verification on the data passed to dataProvider to verify if it has a formEnctype value and if it is set to a 'multipart' value. If so, then it encodes the values to FormData instead of JSON. This will allow posting to endpoints expecting multipart/form-data.

Additionally:
 - Always add tests and ensure they pass. - Check
 - Never break backward compatibility (see https://symfony.com/bc). - Check
 - Bug fixes must be submitted against the current stable version branch. - Check
 - Features and deprecations must be submitted against main branch. - Check
 - Legacy code removals go to the main branch. - Check
 - Update CHANGELOG.md file. - Check
 - Follow the [Conventional Commits specification](https://www.conventionalcommits.org/).
-->

ericovasconcelos commented:
Hi @alanpoulain,

I´ve used 'formEnctype' and 'multipart' as the keywords to force the multipart enctype. 

The HTML specs uses:
- 'formenctype' (in lowercase) and "multipart/form-data" keyword. (https://html.spec.whatwg.org/#dom-fs-formenctype)

The API-Platform uses:
- 'multipart' and 'multipart/form-data' as keywords (https://api-platform.com/docs/core/file-upload).

Do you think it would be better to use one of the above keywords instead of 'formEnctype' and 'multipart'? It might make more sense to be in compliant with one of the standards. I was thinking that it might make more sense to keep the formEnctype and use 'multipart/form-data'. Because formEnctype is the correct IDL attribute name and 'multipart/form-data' is also compliant with HTML.  I guess that since it is running in the form processing it make sense to stick with the HTML naming. 

What do you think?

ericovasconcelos commented:
Hi @alanpoulain,

I´ve also added the update operation support:

- Added tests for update operation without overriding enctype and HTTP verb to POST;
- Added tests to cover the enctype and HTTP verd overriding, using 'formEnctype' and 'formMethod' keywords via data param.

Implemented the code for those scenarios on dataProvider. All tests are green. No BC.

Please see if it is ok.

alanpoulain commented on src/hydra/dataProvider.js:
```diff
@@ -179,8 +179,19 @@ export default (
         Object.values(element).some((value) => value instanceof File);
+      const containsMultipartOverride = (data) => {
+        if (data.formEnctype && data.formEnctype === 'multipart') {
+          delete data.formEnctype;
```
Given the name of this function, it should be pure (not mutating `data`).

ericovasconcelos replied:
Perfect, tks.

Done.

alanpoulain commented on src/hydra/dataProvider.js:
```diff
@@ -334,11 +345,23 @@ export default (
+            params.data.formMethod &&
+            params.data.formMethod === 'POST'
+          ) {
+            delete params.data.formMethod;
```
Same comment as above.

ericovasconcelos replied:
Sure. Tks.

Done.

alanpoulain commented on src/hydra/dataProvider.js:
```diff
@@ -334,11 +345,23 @@ export default (
       case UPDATE:
+        const getUpdateHttpVerb = (params) => {
+          if (
+            params.data &&
```
Couldn't we use the optional chaining here (Babel should probably transpile it correctly)?

ericovasconcelos replied:
Great tip. Done.

alanpoulain commented on src/hydra/dataProvider.test.js:
```diff
@@ -218,4 +218,117 @@ describe('Transform a React Admin request to an Hydra request', () => {
+          formEnctype: 'multipart',
+        },
+      })
+      .then(() => {
```
Why not using async/await syntax?

ericovasconcelos replied:
Sure. Updated tests to use the async/await syntax (also ajusted previous tests on dataProvider.test.js to comply).

Tks for the feedback.

ericovasconcelos replied:
Done.

alanpoulain commented:
> Do you think it would be better to use one of the above keywords instead of 'formEnctype' and 'multipart'? It might make more sense to be compliant with one of the standards. I was thinking that it might make more sense to keep the formEnctype and use 'multipart/form-data'. Because formEnctype is the correct IDL attribute name and 'multipart/form-data' is also compliant with HTML. Since it is running in the form processing on the client-side, it makes sense to stick with the HTML naming.
>
> What do you think?

Using `multipart/form-data` seems a good idea :slightly_smiling_face:

ericovasconcelos commented:
> > Do you think it would be better to use one of the above keywords instead of 'formEnctype' and 'multipart'? It might make more sense to be compliant with one of the standards. I was thinking that it might make more sense to keep the formEnctype and use 'multipart/form-data'. Because formEnctype is the correct IDL attribute name and 'multipart/form-data' is also compliant with HTML. Since it is running in the form processing on the client-side, it makes sense to stick with the HTML naming.
> > What do you think?
> 
> Using `multipart/form-data` seems a good idea 🙂

Dons ;)

alanpoulain commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## 2.6.5
+
+* Accept 'formEnctype: multipart' in data passed to data provider to force multipart encoding.
```
```suggestion
* Accept 'formEnctype: multipart/form-data' in data passed to data provider to force multipart encoding.
```

alanpoulain commented on src/hydra/dataProvider.js:
```diff
@@ -179,11 +179,21 @@ export default (
         Object.values(element).some((value) => value instanceof File);
-
-      if (!values.some((value) => containFile(value))) {
+      const containsMultipartOverride = (data) => {
+        return data.formEnctype === 'multipart/form-data' ? true : false;
+      };
```
```suggestion
      const containsMultipartOverride = (data) => data.formEnctype === 'multipart/form-data';
```

alanpoulain commented on src/hydra/dataProvider.js:
```diff
@@ -334,11 +344,14 @@ export default (
         });
 
       case UPDATE:
+        const updateHttpVerb =
+          params.data?.formMethod === 'POST' ? 'POST' : 'PUT';
```
```suggestion
        const updateHttpMethod =
          params.data?.formMethod ?? 'PUT';
```

alanpoulain commented:
After some thinking, I'm not sure about the terms `formEnctype` and `formMethod`.
Since it's the data provider receiving the information, it should not be related to a "form".
I would prefer to use for instance `mediaType` and `httpMethod`.

Also, I don't really like having to delete this extra information from `data`. Ideally they should not be transferred like this but since we don't have a choice, we should make sure to not be specific to the hydra data provider.
That's why we should use a unique "extra" property in `data` used only for passing extra information to the data provider.
For instance:
```js
const transform = data => ({
    ...data, 
    extraInformation: {
        mediaType: 'multipart/form-data',
        httpMethod: 'POST',
    },
})
```
This `extraInformation` key would be then deleted once in each data provider.

And IMO we should go even further: `mediaType` and `httpMethod` are too related to the chosen data provider. We should be less specific.
For instance something like this:
```js
const transform = data => ({
    ...data, 
    extraInformation: {
        hasFile: true,
    },
})
```
And the hydra data provider then knows it has to use a multipart/form-data media type and always use the POST method.
WDYT?

ericovasconcelos commented:
> After some thinking, I'm not sure about the terms `formEnctype` and `formMethod`.
> Since it's the data provider receiving the information, it should not be related to a "form".
> I would prefer to use for instance `mediaType` and `httpMethod`.
> 
> Also, I don't really like having to delete this extra information from `data`. Ideally they should not be transferred like this but since we don't have a choice, we should make sure to not be specific to the hydra data provider.
> That's why we should use a unique "extra" property in `data` used only for passing extra information to the data provider.
> For instance:
> 
> ```js
> const transform = data => ({
>     ...data, 
>     extraInformation: {
>         mediaType: 'multipart/form-data',
>         httpMethod: 'POST',
>     },
> })
> ```
> 
> This `extraInformation` key would be then deleted once in each data provider.
> 
> And IMO we should go even further: `mediaType` and `httpMethod` are too related to the chosen data provider. We should be less specific.
> For instance something like this:
> 
> ```js
> const transform = data => ({
>     ...data, 
>     extraInformation: {
>         hasFile: true,
>     },
> })
> ```
> 
> And the hydra data provider then knows it has to use a multipart/form-data media type and always use the POST method.
> WDYT?

I like the idea. Makes very straightforward and clear to anyone that will use. I would just use hasFileField instead of just hasFile to make more clear that we just need the field to be present, WDYT?

May I will proceed to update the code with your suggestions?

alanpoulain commented:
Sure, `hasFileField` is alright.
> May I will proceed to update the code with your suggestions?
Yes you can! Sorry for changing everything after you have already worked on it.

ericovasconcelos commented:
Hi @alanpoulain,

I think the code looks better after your suggestions. Thanks for the feedback. I´ve done the changes and added more tests to make sure the provider totally ignores the extraInformation key (even if there is no hasFileField set to true).

alanpoulain commented:
Hello @ericovasconcelos. I've reworked a little your PR to delete the `extraInformation` key sooner.
I've also removed the last tests, I think it is already tested with the previous ones.
Are you OK with the changes?

ericovasconcelos reviewed (commented):
Great.

ericovasconcelos commented:
> Hello @ericovasconcelos. I've reworked a little your PR to delete the `extraInformation` key sooner.
> I've also removed the last tests, I think it is already tested with the previous ones.
> Are you OK with the changes?

Sure. Great job on dealing with extraInformation sooner on the code!

ericovasconcelos commented:
Hi @alanpoulain, is there any work left to do on this PR?

ericovasconcelos commented:
Hi @alanpoulain , is there any work left to do on this pull request?

alanpoulain commented:
Sorry for the delay, thank you @ericovasconcelos.
